libyul/optimiser/ConstantOptimiser.cpp

```diff
@@ -0,0 +1,181 @@
+#include <libyul/Utilities.h>
+#include <libyul/AsmParser.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using Representation = ConstantOptimiser::Representation;
+
+void ConstantOptimiser::visit(Expression& _e)
+{
+	if (_e.type() == typeid(Literal))
+	{
+		Literal const& literal = boost::get<Literal>(_e);
+		if (literal.kind != LiteralKind::Number)
+			return;
+		u256 value = valueOfLiteral(literal);
+		if (value < 0x10000)
+			return;
+
+		if (
+			Expression const* repr =
+			RepresentationFinder(m_evmVersion, m_meter, locationOf(_e), m_cache)
+			.tryFindRepresentation(value)
+		)
+			_e = ASTCopier{}.translate(*repr);
+	}
+	else
+		ASTModifier::visit(_e);
+}
+
+Expression const* RepresentationFinder::tryFindRepresentation(dev::u256 const& _value)
+{
+	if (_value < 0x10000)
+		return nullptr;
+
+	Representation repr = findRepresentation(_value);
+	if (repr.expression->type() == typeid(Literal))
+		return nullptr;
+	else
+		return repr.expression.get();
+}
+
+Representation RepresentationFinder::findRepresentation(dev::u256 const& _value)
+{
+	if (m_cache.count(_value))
+		return m_cache.at(_value);
+
+	Representation routine;
+	if (_value <= 0x10000)
+		routine = min(move(routine), represent(_value));
+	else if (dev::bytesRequired(~_value) < dev::bytesRequired(_value))
+		// Negated is shorter to represent
+		routine = min(move(routine), represent("not", findRepresentation(~_value)));
+	else
+	{
+		// Decompose value into a * 2**k + b where abs(b) << 2**k
+		// Is not always better, try literal and decomposition method.
+		routine = represent(_value);
+
+		for (unsigned bits = 255; bits > 8 && m_maxSteps > 0; --bits)
+		{
+			unsigned gapDetector = unsigned((_value >> (bits - 8)) & 0x1ff);
+			if (gapDetector != 0xff && gapDetector != 0x100)
+				continue;
+
+			u256 powerOfTwo = u256(1) << bits;
+			u256 upperPart = _value >> bits;
+			bigint lowerPart = _value & (powerOfTwo - 1);
+			if ((powerOfTwo - lowerPart) < lowerPart)
+			{
+				lowerPart = lowerPart - powerOfTwo; // make it negative
+				upperPart++;
+			}
+			if (upperPart == 0)
+				continue;
+			if (abs(lowerPart) >= (powerOfTwo >> 8))
+				continue;
+			Representation newRoutine;
+			if (m_evmVersion.hasBitwiseShifting())
+				newRoutine = represent("shl", represent(bits), findRepresentation(upperPart));
+			else
+			{
+				newRoutine = represent("exp", represent(2), represent(bits));
+				if (upperPart != 1)
+					newRoutine = represent("mul", findRepresentation(upperPart), newRoutine);
+			}
+
+			if (newRoutine.cost >= routine.cost)
+				continue;
+
+			if (lowerPart > 0)
+				newRoutine = represent("add", newRoutine, findRepresentation(u256(abs(lowerPart))));
+			else if (lowerPart < 0)
+				newRoutine = represent("sub", newRoutine, findRepresentation(u256(abs(lowerPart))));
+
+			if (m_maxSteps > 0)
+				m_maxSteps--;
+			routine = min(move(routine), move(newRoutine));
+		}
+	}
+	m_cache[_value] = routine;
```

Comment:
Looks like this could move routine in the cache and then return a pointer to it...

Reply:
That is if ``Representation`` contained a ``unique_ptr``... ``findRepresentation`` could return a plain pointer to ``Representation``... ``represent`` should still always return a ``Representation`` by value... maybe I'm missing some complication there, but to me it looks like that should be fine...

Reply:
findRepresentation still needs to return a Representation because we need the cost for recursive calls.

Reply:
Yeah? The recursive calls only pass them along to some ``represent``, so we could just dereference the plain pointer... I don't see a call to ``findRepresentation`` for which that wouldn't work...

Reply:
I meant a plain pointer to the ``Representation`` object in the cache, maybe that wasn't clear...

Reply:
``findRepresentation`` only returns things that will continue to be in the cache... it could in fact even return a reference I think :-).

Reply:
Yes, that makes sense!

Reply:
If it returns a reference then a unique_ptr doesn't work.

Reply:
Ah sorry, it still does!